### PR TITLE
[SPIKE] [CORL-1012] Twitter and YouTube embeds

### PR DESCRIPTION
**Pull request by nick-funk**

### CODE REVIEW

File: src/core/server/graph/schema/schema.graphql [hunk 1]
```diff
@@ -1240,6 +1240,22 @@ type StoryMessageBox {
...
+type EmbedLinksConfiguration {
+  """
+  twitterEnabled when true will enable twitter link embeds in the stream.
+  """
+  twitterEnabled: Boolean!
+
+  """
+  youtubeEnabled when true will enable youtube link embeds in the stream.
+  """
+  youtubeEnabled: Boolean!
```
Comment (wyattjoh) [inline src/core/server/graph/schema/schema.graphql]: Suggestion: Rename fields to `twitter: Boolean!` and `youtube: Boolean!`. The "Enabled" part of the property can be inferred from the type.

File: src/core/server/app/handlers/api/oEmbed/oEmbed.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return `https://publish.twitter.com/oembed?url=${url}`;
+  }
+  if (type === "youtube") {
+    return `https://www.youtube.com/oembed?url=${url}`;
+  }
+
+  return "";
+};
+
+export const oEmbedHandler = (): RequestHandler => {
```
Comment (wyattjoh) [inline src/core/server/app/handlers/api/oEmbed/oEmbed.ts]: Handlers should be reflective of their mounting points. Coral enforces a strict lowercase on URL's, this should be renamed to:

```suggestion
export const oembedHandler = (): RequestHandler => {
```

File: src/core/server/app/handlers/api/oEmbed/oEmbed.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+const fetchURL = (url: string, type: string) => {
+  if (type === "twitter") {
+    return `https://publish.twitter.com/oembed?url=${url}`;
+  }
+  if (type === "youtube") {
+    return `https://www.youtube.com/oembed?url=${url}`;
+  }
+
+  return "";
+};
```
Comment (wyattjoh) [inline src/core/server/app/handlers/api/oEmbed/oEmbed.ts]: All URL parameters should be encoded with `encodeURIComponent`. While this is handled in the handler, it should instead be moved here to ensure that it's protected.

File: src/core/server/graph/schema/schema.graphql [hunk 4]
```diff
@@ -1240,6 +1240,22 @@ type StoryMessageBox {
   content: String
 }
 
+################################################################################
+## Embed Links
+################################################################################
+
+type EmbedLinksConfiguration {
```
Comment (wyattjoh) [inline src/core/server/graph/schema/schema.graphql]: As this can likely be the integration point for other embedded content, suggest renaming:

```suggestion
type EmbedConfiguration {
```

File: src/core/server/graph/schema/schema.graphql [hunk 5]
```diff
@@ -2461,6 +2483,30 @@ type CommentRevisionMetadata {
...
+  """
+  YOUTUBE
+
+  """
+  TWITTER represents a twitter link source.
+  """
+  TWITTER
+}
+
+type CommentEmbedLink {
```
Comment (wyattjoh) [inline src/core/server/graph/schema/schema.graphql]: Same suggestion as above:

```suggestion
type CommentEmbed {
```

File: src/core/server/graph/schema/schema.graphql [hunk 6]
```diff
@@ -2485,6 +2531,12 @@ type CommentRevision {
   """
   body: String
 
+  """
+  embedLinks is the collection of links that were identified as supported
+  embedded content in the comment body
+  """
+  embedLinks: [CommentEmbedLink!]
```
Comment (wyattjoh) [inline src/core/server/graph/schema/schema.graphql]:
```suggestion
  embed: [CommentEmbed!]
```

File: src/core/server/app/handlers/api/oEmbed/oEmbed.ts (hunk 3 above)
Comment (nick-funk) [reply src/core/server/app/handlers/api/oEmbed/oEmbed.ts]: I have done this, but had to remove the previous encoding on the url from the request. Reason being, you cannot call `encodeURIComponent` on a url twice... it will garble the url string.

**Comment by nick-funk** [comment]
Closing as @tessalt has migrated this into another branch that combines GIF + YouTube + Twitter together into one consolidated project.

tldr; the Twitter/YouTube spike is over.
